File: android/src/main/java/com/compdfkit/flutter/compdfkit_flutter/utils/CPDFDocumentInfoUtil.java
/*
 * Copyright © 2014-2026 PDF Technologies, Inc. All Rights Reserved.
 *
 * THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
 * AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE ComPDFKit LICENSE AGREEMENT.
 * UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
 * This notice may not be removed from this file.
 *
 */

package com.compdfkit.flutter.compdfkit_flutter.utils;

import android.util.Log;
import com.compdfkit.core.document.CPDFDocument;
import com.compdfkit.core.document.CPDFDocumentPermissionInfo;
import com.compdfkit.core.document.CPDFInfo;
import com.compdfkit.tools.common.utils.date.CDateUtil;
import java.util.HashMap;
import java.util.Map;

public class CPDFDocumentInfoUtil {

  public static Map<String, Object> getDocumentInfo(CPDFDocument document) {
    Map<String, Object> documentInfo = new HashMap<>();
    CPDFInfo info = document.getInfo();
    try {
      documentInfo.put("title", info.getTitle());
      documentInfo.put("author", info.getAuthor());
      documentInfo.put("subject", info.getSubject());
      documentInfo.put("keywords", info.getKeywords());
      documentInfo.put("creator", info.getCreator());
      documentInfo.put("producer", info.getProducer());
      long createDateTimes = CAppUtils.toTimes(CDateUtil.transformPDFDate(info.getCreationDate()));
      if (createDateTimes != 0L) {
        documentInfo.put("creationDate", createDateTimes);
      }
      long modificationDateTimes = CAppUtils.toTimes(CDateUtil.transformPDFDate(info.getModificationDate()));
      if (modificationDateTimes != 0L) {
        documentInfo.put("modificationDate", modificationDateTimes);
      }
      return documentInfo;
    } catch (Exception e) {
      return documentInfo;
    }
  }

  public static Map<String, Object> getPermissionsInfo(CPDFDocument document) {
    Map<String, Object> map = new HashMap<>();
    try {
      CPDFDocumentPermissionInfo permissionInfo = document.getPermissionsInfo();
      map.put("allowsPrinting", permissionInfo.isAllowsPrinting());
      map.put("allowsHighQualityPrinting", permissionInfo.isAllowsHighQualityPrinting());
      map.put("allowsCopying", permissionInfo.isAllowsCopying());
      map.put("allowsDocumentChanges", permissionInfo.isAllowsDocumentChanges());
      map.put("allowsDocumentAssembly", permissionInfo.isAllowsDocumentAssembly());
      map.put("allowsCommenting", permissionInfo.isAllowsCommenting());
      map.put("allowsFormFieldEntry", permissionInfo.isAllowsFormFieldEntry());
      return map;
    } catch (Exception e) {
      return map;
    }
  }
}

File: ios/Classes/reader/CPDFDocumentInfoUtil.swift
//  CPDFDocumentInfoUtil.swift
//  compdfkit_flutter
//
//  Copyright © 2014-2026 PDF Technologies, Inc. All Rights Reserved.
//
//  THIS SOURCE CODE AND ANY ACCOMPANYING DOCUMENTATION ARE PROTECTED BY INTERNATIONAL COPYRIGHT LAW
//  AND MAY NOT BE RESOLD OR REDISTRIBUTED. USAGE IS BOUND TO THE ComPDFKit LICENSE AGREEMENT.
//  UNAUTHORIZED REPRODUCTION OR DISTRIBUTION IS SUBJECT TO CIVIL AND CRIMINAL PENALTIES.
//  This notice may not be removed from this file.
//


import Foundation
import ComPDFKit
import Flutter
import ComPDFKit_Tools


public class CPDFDocumentInfoUtil {
    
    
    static func getDocumentInfo(from document: CPDFDocument?) -> [String: Any] {
        var documentInfo: [String: Any] = [:]
        
        guard let document = document,
              let documentAttributes = document.documentAttributes() else {
            return documentInfo
        }
        
        documentInfo["title"] = documentAttributes[.titleAttribute] as? String
        documentInfo["author"] = documentAttributes[.authorAttribute] as? String
        documentInfo["subject"] = documentAttributes[.subjectAttribute] as? String
        documentInfo["creator"] = documentAttributes[.creatorAttribute] as? String
        documentInfo["keywords"] = documentAttributes[.keywordsAttribute] as? String
        
        if let creationDateString = documentAttributes[.creationDateAttribute] as? String {
            documentInfo["creationDate"] = convertPDFDateToTimestamp(creationDateString)
        }
        
        if let modificationDateString = documentAttributes[.modificationDateAttribute] as? String {
            documentInfo["modificationDate"] = convertPDFDateToTimestamp(modificationDateString)
        }
        
        return documentInfo
    }
    
    
    private static func convertPDFDateToTimestamp(_ pdfDateString: String) -> Int? {
        guard let date = parsePDFDateString(pdfDateString) else {
            return nil
        }
        return Int(date.timeIntervalSince1970 * 1000)
    }
    
    private static func parsePDFDateString(_ pdfDateString: String) -> Date? {
        let formattedDateString = formatPDFDateString(pdfDateString)
        
        guard !formattedDateString.isEmpty else {
            return nil
        }
        
        return parseFormattedDateString(formattedDateString)
    }
    
    private static func formatPDFDateString(_ pdfDateString: String) -> String {
        guard pdfDateString.count >= 16 else {
            return ""
        }
        
        let startIndex = pdfDateString.index(pdfDateString.startIndex, offsetBy: 2)
        
        let year = extractSubstring(from: pdfDateString, startIndex: startIndex, offset: 0, length: 4)
        let month = extractSubstring(from: pdfDateString, startIndex: startIndex, offset: 4, length: 2)
        let day = extractSubstring(from: pdfDateString, startIndex: startIndex, offset: 6, length: 2)
        let hour = extractSubstring(from: pdfDateString, startIndex: startIndex, offset: 8, length: 2)
        let minute = extractSubstring(from: pdfDateString, startIndex: startIndex, offset: 10, length: 2)
        let second = extractSubstring(from: pdfDateString, startIndex: startIndex, offset: 12, length: 2)
        let formattedString = "\(year)-\(month)-\(day) \(hour):\(minute):\(second)"
        return formattedString
    }
    

    private static func extractSubstring(from string: String,
                                         startIndex: String.Index,
                                         offset: Int,
                                         length: Int) -> String {
        let rangeStart = string.index(startIndex, offsetBy: offset)
        let rangeEnd = string.index(rangeStart, offsetBy: length)
        return String(string[rangeStart..<rangeEnd])
    }

    
    private static func parseFormattedDateString(_ dateString: String) -> Date? {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        return formatter.date(from: dateString)
    }
    
    static func getPermissionsInfo(document : CPDFDocument?) -> [String: Any] {
        var permissionsInfo: [String: Any] = [:]
        if(document == nil){
            return permissionsInfo;
        }
        permissionsInfo["allowsPrinting"] = document!.allowsPrinting
        permissionsInfo["allowsHighQualityPrinting"] = document!.allowsHighQualityPrinting
        permissionsInfo["allowsCopying"] = document!.allowsCopying
        permissionsInfo["allowsDocumentChanges"] = document!.allowsDocumentChanges
        permissionsInfo["allowsDocumentAssembly"] = document!.allowsDocumentAssembly
        permissionsInfo["allowsCommenting"] = document!.allowsCommenting
        permissionsInfo["allowsFormFieldEntry"] = document!.allowsFormFieldEntry
        return permissionsInfo
    }
    
}
